Task1_Ivan/PtrStringList.cpp

```diff
@@ -0,0 +1,249 @@
+    {
+        if (!list)
+            return 0;
+
+        char** currentNode = list;
+        size_t i = 1;                           // Head not null -> size >= 1
+        while (StringListForward(&currentNode)) // Traverse list until reached tail, increment each iteration
+            ++i;
+        return i;
+    }
+
+    int slst::StringListIndexOf(char** node, const char* str)
+    {
+        int index = 0;
+        while (node)
+        {
+            char* nextValue = *node;
+            if (util::StringsEqual(nextValue, str))
+                return index;
+            ++index;
+            StringListForward(&node);
+        }
+        return -1;
+    }
+
+    void slst::StringListRemoveDuplicates(char** currentNode)
+    {
+        while (currentNode /* != LIST_END */)
+        {
+            char** next = StringListGetNext(currentNode);
+            StringListRemoveOrClear(&next, currentNode[0]);  // Next might be a duplicate, and be deleted,
+            currentNode[1] = reinterpret_cast<char*>(next);  // which will make currentNode.next a wild pointer, so reassign
+            currentNode = next;
+        }
+    }
+
+    void slst::StringListReplaceInStrings(char** node, char* before, char* after)
+    {
+        if (!node || (before == after) || !before || !after || (strcmp(before, after) == 0))
+            return;
+
+        do
+        {
+            if (node[0])
+            {
+                char* needsReplace = strstr(node[0], before);
+                if (needsReplace)
+                {
+                    char* res = util::StringReplace(node[0], before, after);
+                    free(node[0]);
+                    node[0] = res;
+                }
+            }
+        
+            StringListForward(&node);
+        } while (node  /* != LIST_END */);
+    }
+
+    void slst::StringListSort(char** list)
+    { // Uses selection sort
+        char** currentNode = list;
+        while (currentNode  /* != LIST_END */)       // Traverse the List 
+        {
+            char** maxValNode = currentNode;
+            char** unsortedNode = StringListGetNext(currentNode);
+
+            while (unsortedNode)                     // Traverse the unsorted sublist 
+            {
+                if ((maxValNode[0] != unsortedNode[0]) &&
+                    (maxValNode[0] && unsortedNode[0]) &&  // Both not null for strcmp
+                    (strcmp(maxValNode[0], unsortedNode[0]) > 0))
+                    maxValNode = unsortedNode;
+
+                StringListForward(&unsortedNode);
+            }
+
+            char* dataTemp = currentNode[0];         // Swap
+            currentNode[0] = maxValNode[0];
+            maxValNode[0] = dataTemp;
+
+            StringListForward(&currentNode);
+        }
+    }
+
+    /* Helpers */
+#pragma region HelperFunctions
+    char** slst::StringListAllocateNode()
+    {
+        char** node = util::allocate_memory<char*>(NODE_SIZE);  // Allocate space for 2 pointers to char*

```

Comment:
I does not need to use templates for current task

Reply:
Of course, i will remove it.